File: mps-m/Mp.Mod.Calculation/SampleRatePS.cs
//    MeaProcess - Meaurement and Automation framework.
//    Copyright (C) 2015  Laurentiu-Gheorghe Crisan
//
//    This program is free software: you can redistribute it and/or modify
//    it under the terms of the GNU General Public License as published by
//    the Free Software Foundation, either version 3 of the License, or
//    (at your option) any later version.
//
//    This program is distributed in the hope that it will be useful,
//    but WITHOUT ANY WARRANTY; without even the implied warranty of
//    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//    GNU General Public License for more details.
//
//    You should have received a copy of the GNU General Public License
//    along with this program.  If not, see <http://www.gnu.org/licenses/>.

using System;
using System.Drawing;
using System.Windows.Forms;
using System.Xml;
using Mp.Scheme.Sdk;
using Mp.Utils;

namespace Mp.Mod.Calculation
{
    internal class SampleRatePS : WorkPS
    {
        public SampleRatePS()
        {
            base.Type = "Mp.Calculation.PS.SampleRate";
            base.Text = StringResource.SampleRate;
            base.Group = StringResource.Calculation;
            base.Symbol = Resource.SampleRateImg;
            base.Icon = Resource.SampleRateIcon;
            base.IsSingleton = false;
        }

        public override string RuntimeModule
        {
            get { return "mps-calculation"; }
        }

        public override void OnLoadResources()
        {
            base.Text = StringResource.SampleRate;
            base.Group = StringResource.Calculation;
        }

        public override void OnDefaultInit()
        {
            base.OnDefaultInit();

            //Create the data out port.
            Port port = new Port(new Point(_rectangle.Right + PortWidth, (int)(_rectangle.Top + PortTopOffset)), "Mp.Port.Out", false, true);
            port.SignalList = Document.CreateSignalList();
            InitMenuForPort(port);
            AddPort(port);

            //Data in port. 
            port = new Port(new Point(_rectangle.Left - PortWidth, (int)(_rectangle.Top + PortTopOffset)), "Mp.Port.In", true, false);
            AddPort(port);
        }

        public override void OnLoadXml()
        {
            base.OnLoadXml();
            InitMenuForPort(OutputPorts[0]);
        }

        public override string Description
        {
            get
            {
                return StringResource.SampleRateDescription;
            }
        }

        private void InitMenuForPort(Port port)
        {
            //Create the context menu.
            port.ContextMenuStrip = new ContextMenuStrip();
            port.ContextMenuStrip.Tag = port;

            ToolStripMenuItem menuItem = new ToolStripMenuItem(StringResource.MenuProperties);

            menuItem.Font = new System.Drawing.Font("Tahoma", 8.25F, System.Drawing.FontStyle.Bold);
            menuItem.Click += new System.EventHandler(this.OnPropertyDataPort);
            port.ContextMenuStrip.Items.Add(menuItem);
        }

        protected override void OnDocumentChanged()
        {
            if (InputPorts.Count == 0)
                return;

            XmlElement xmlOutSignalList = OutputPorts[0].SignalList;

            if (xmlOutSignalList == null)
                return;

            Port port = InputPorts[0];
            //Update the signals ids
            for (int i = 0; i < xmlOutSignalList.ChildNodes.Count; ++i)
            {
                XmlElement xmlOutSignal = (XmlElement)xmlOutSignalList.ChildNodes[i];


                uint sigId = (uint)XmlHelper.GetParamNumber(xmlOutSignal, "inSignal");

                XmlElement xmlSig = Document.GetXmlObjectById(sigId);

                if (xmlSig == null || port.SignalList == null)
                {
                    xmlOutSignalList.RemoveChild(xmlOutSignal);
                    i--;
                    continue;
                }

                bool found = false;
                foreach (XmlElement xmlPortSig in port.SignalList.ChildNodes)
                {
                    XmlElement xmlSignal = xmlPortSig;

                    if (XmlHelper.GetObjectID(xmlPortSig) == 0)
                        xmlSignal = Document.GetXmlObjectById(Convert.ToUInt32(xmlPortSig.InnerText));

                    if (xmlSignal == xmlSig)
                    {
                        found = true;
                        break;
                    }
                }

                if (!found)
                {
                    xmlOutSignalList.RemoveChild(xmlOutSignal);
                    i--;
                    continue;
                }
                else
                {
                    //Update the output signal
                    double min = XmlHelper.GetParamDouble(xmlSig, "physMin");
                    double max = XmlHelper.GetParamDouble(xmlSig, "physMax");
                    XmlHelper.SetParamDouble(xmlOutSignal, "physMin", "double", min);
                    XmlHelper.SetParamDouble(xmlOutSignal, "physMax", "double", max);
                    XmlHelper.SetParam(xmlOutSignal, "parameters", "string", XmlHelper.GetParam(xmlSig, "parameters"));
                }
            }
        }

        public override void OnPortDoubleClick(Port port)
        {
            base.OnPortDoubleClick(port);

            if (port.IsInput)
                return;

            OnPropertyDataPort(null, null);
        }

        protected void OnPropertyDataPort(object sender, EventArgs e)
        {
            Port port = OutputPorts[0];
            SignalViewDlg dlg = new SignalViewDlg(port.SignalList, Document);
            dlg.ShowDialog();
        }

        public override void OnHelpRequested()
        {
            Document.ShowHelp(this.Site, 1440);
        }

        protected override void OnProperties(object sender, EventArgs e)
        {
            SampleRateDlg dlg = new SampleRateDlg(XmlRep, Document, InputPorts[0].SignalList, OutputPorts[0].SignalList);

            if (dlg.ShowDialog() == DialogResult.OK)
                this.Text = XmlHelper.GetParam(XmlRep, "name");             
        }
    }
}

File: mps-n/src/mps-calculation/SampleRatePS.cpp
/*
 * Copyright (C) 2015 by Laurentiu-Gheorghe Crisan
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * As a special exception, you may use this file as part of a free
 * software library without restriction. Specifically, if other files
 * instantiate templates or use macros or inline functions from this
 * file, or you compile this file and link it with other files to
 * produce an executable, this file does not by itself cause the
 * resulting executable to be covered by the GNU General Public
 * License. This exception does not however invalidate any other
 * reasons why the executable file might be covered by the GNU Library
 * General Public License.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
 */
#include "SampleRatePS.h"
#include <mps/core/SignalList.h>
#include "SampleRateSignal.h"

namespace mps{
namespace calculation{

SampleRatePS::SampleRatePS()
{
}

SampleRatePS::~SampleRatePS()
{
}

void SampleRatePS::onInitInstance()
{
    FiFoSynchSourcePS::onInitInstance();

    const mps::core::Port* outPort = _outputPorts->at(0);

    for( Pt::uint32_t i = 0; i < outPort->signalList()->size(); ++i)
    {
        SampleRateSignal* signal = (SampleRateSignal*) outPort->signalList()->at(i);
        signal->setIndex(i);
        InSig2OutSigIt it = _inSig2OutSig.find(signal->inSignal());

        if( it == _inSig2OutSig.end())
        {
            std::vector<SampleRateSignal*> outSignals;

            outSignals.push_back(signal);

            std::pair<Pt::uint32_t, std::vector<SampleRateSignal*> > pair(signal->inSignal(), outSignals);
            _inSig2OutSig.insert(pair);
        }
        else
        {
            it->second.push_back(signal);
        }
    }

    _sampleIncrement.resize(outPort->signalList()->size());
}

void SampleRatePS::onStart()
{
    for( Pt::uint32_t i = 0; i < _sampleIncrement.size(); ++i)
        _sampleIncrement[i] = 0;

    FiFoSynchSourcePS::onStart();
}

void SampleRatePS::onUpdateDataValue(Pt::uint32_t noOfRecords, Pt::uint32_t sourceIdx, const mps::core::Port* port, const Pt::uint8_t* data)
{
    const mps::core::Sources& sources = port->sources();
    const std::vector<mps::core::Signal*>& source = sources[sourceIdx];
    const Pt::uint32_t inRecordSize = port->sourceDataSize(sourceIdx);

    for( Pt::uint32_t i = 0; i < source.size(); ++i)
    {
        const mps::core::Signal* signal = source[i];
        InSig2OutSigIt it = _inSig2OutSig.find(signal->signalID());

        if( it == _inSig2OutSig.end())
            continue;

        const Pt::uint32_t sigOffset = port->signalOffsetInSource(sourceIdx, i);

        for(Pt::uint32_t rec = 0; rec < noOfRecords; ++rec)
        {
            const Pt::uint8_t* pvalue = &data[rec * inRecordSize + sigOffset];

            for( Pt::uint32_t k = 0; k < it->second.size(); ++k)
            {
                const SampleRateSignal* outSignal = it->second[k];
                _sampleIncrement[outSignal->index()] += (outSignal->sampleRate() /signal->sampleRate() );

                int z = 0;

                for( ; z < (int)_sampleIncrement[outSignal->index()]; ++z)
                    putValue(outSignal->index(),0,pvalue);
                    
                _sampleIncrement[outSignal->index()] -= z;
            }
        }
    }
}

}}
